### ISSUE TITLE
AMQ Queues don't always show as Queues

### ISSUE THREAD

**Issue by geerzo** [hawtio/hawtio#1921]
We have a periodic problem where queues won't register as full up queues. As you can see in the image the icons are different and when you click on them they aren't treated as queues (no attributes, operations, etc). We are running 1.4.25 but I couldn't find any known issues for this.

![pid_11789_activemq](https://cloud.githubusercontent.com/assets/328937/6557628/a85cb5f6-c633-11e4-82af-f51fc6d0128c.png)

**Comment by gashcrumb** [comment]
Try looking at the queue on the JMX tab perhaps and see if the attributes can be viewed there.  If possible, take a note of the ObjectName property of the mbean so we can compare it to one of the working queues.  If I recall correctly hawtio figures out if an mbean represents a queue based on the ObjectName, could be that code needs to be adjusted accordingly.  Thanks!

**Comment by geerzo** [comment]
I did try that as well and they aren't recognized on the JMX tab either. We've been running into this issue for a while not but it always affects different queues at different times. I'm not sure if its tied to a Karaf reboot or just changes randomly. It's actually so bad that there are no attributes listed at all, hawtio just treats it like a folder.

**Comment by eschizoid** [comment]
I have the exact same problem that @geerzo. When Karaf starts all the queues are listed properly but as soon as time pass the hawtio console stop showing some queues or just treat them as folders. If I go to the ActiveMQ web console all the queues are listed properly.

**Comment by gtmyers** [comment]
I am also seeing this behavior in karaf-2.4.1, ActiveMQ-5.10.0 and Hawtio-1.4.45.  Queues displaying as folders with no attributes.  The same issue was reported here with no obvious resolution.
https://github.com/hawtio/hawtio/issues/623
I have been using JConsole as a workaround for many months.  Any help would be greatly appreciated.

**Comment by davsclaus** [comment]
It may be that jolokia bug that would limit the number of mbeans returned. Try upgrading jolokia on your AMQ / karaf to latest release.

**Comment by dcruz-interware** [comment]
How do i upgrade jolokia? is there a useful link?

**Comment by davsclaus** [comment]
You can uninstall the old jolokia and install the newer version in karaf, it has a CLI with install and uninstall commands.

**Comment by rlodge** [comment]
This occurs for me all the time, using the latest AMQ (5.12.0) which says it's running the latest Jolokia (1.3.1).

**Comment by davsclaus** [comment]
There is also a limit in hawtio in that older release that limits hawtio to return 500/1000 at most I think. You may need to go to preference and find that setting and set it higher

**Comment by davsclaus** [comment]
There is a new 1.4.54 release of hawtio

**Comment by rlodge** [comment]
1.4.54 doesn't seem to make any difference.

**Comment by davsclaus** [comment]
okay we got from users with 1.4.55 that has a new jolokia version that it fixed a problem for them with missing Camels. So would be good to get this ticket tested with the latest release.

**Comment by rlodge** [comment]
I'm not seeing any difference.  My problem is with ActiveMQ queues / topics and not camel routes in any case.  Does the version of jolokia on the remote end (AMQ) matter?

**Comment by davsclaus** [comment]
yes the queues are also represented in jmx and therefore also jolokia on the server side should be jolokia 1.3.2 in hope of resolving this.

**Comment by rlodge** [comment]
I updated my activemq (5.12) install to jolokia 1.3.2, and it did not seem to help.

**Comment by eschizoid** [comment]
Same here, current setup:
ActiveMQ 5.12.0
Hawtio 1.4.55

**Comment by davsclaus** [comment]
please take a look in regular jconsole and see those mysterious queues what their mbreans are represented from jconsole, and compare them with a working queue. It may be something in the object name or something that could explain the difference.

**Comment by eschizoid** [comment]
@davsclaus After doing some comparison between jvisualvm and hawtio, the number of queues listed in hawtio is much much lower. Attached you will find the some screenshot from both:

![screen shot 2015-11-27 at 11 51 04 pm](https://cloud.githubusercontent.com/assets/8173837/11450337/20116abc-9562-11e5-898b-9cd8662cc58a.png)

![screen shot 2015-11-27 at 11 49 48 pm](https://cloud.githubusercontent.com/assets/8173837/11450339/2525bfc6-9562-11e5-850a-e421a40b24fd.png)

I don't see anything wrong/weird with the name of the queues.

**Comment by davsclaus** [comment]
The traditional web console do not use jolokia, its jsp based.

I suspect its a problem with both jolokia that had an built-in upper cap in the max number of mbeans to return. And a setting in hawtio where you can limit that as well.

That can explain that its "random" number of queues that are not shown, as that upper cap is hit and no more data is returned.

**Comment by atooni** [comment]
Just came across this thread. We are seeing exactly the same behavior with ActiveMQ 5.13.0 and Hawtio 1.4.61.

**Comment by atooni** [comment]
I spend a bit of time to nail down when and why this happens. We do have an OSGI container that hosts ActiveMQ 5.13.0, Camel 2.16.1 and HawtIO 1.4.61 (with Jolokia 1.3.3). 

On some machines we can see the full set of queues of our application without any problems, on other machines we experience the behavior described here right from the beginning. The machines that work fine may enter the "weird" state after an undetermined amount of time. Once a machine has reached that state, it will never turn back to normal. Even a complete reinstallation, deleting the browser cache will not bring it back to a working state. 

Reading what has been said before, I can provoke the weird state by restricting the maximum collection size returned from Jolokia to a ridiculous low value (such as 30). Resetting the maximum size to 5000 will restore the UI to it's normal working state. 

This having said, changing the parameters for a UI that is currently "weird" (i.e. increase the maximum collection size) does not repair the UI. However - the number of Queues displayed as folders changes sometimes after modifying those parameters. 

I have been through the code and have seen there a number of hard coded timeout values (my guess is that the UI waits a certain time on the returned MBeans and then refreshes the UI. I was wondering whether it is worth making those timeouts configurable.

**Comment by atooni** [comment]
In one of our affected containers we were able to see the full tree correctly after increasing the maximum collection size to 35000. However, displaying the tree took about 15 minutes.

**Comment by davsclaus** [comment]
yeah @grgrzybek is working on a performance improvement to load the tree.

**Comment by grgrzybek** [comment]
@atooni yes - I hope to fix this by Friday.
There's fix on Karaf/RBAC side as well as in `hawtio-osgi-jmx` bundle and hawtio Angular app itself. Do you use hawtio standalone in Karaf? in Fabric8? in Fuse?

**Comment by atooni** [comment]
@grgrzybek - That sounds very good. We are using hawt.io in our own OSGI container (https://github.com/woq-blended/blended) Essentially this is a Scala based container which has ActiveMQ 5.13.0 as a JMS layer, Camel 2.16.0 as a routing engine and the Scala stack for REST fule web services.

**Comment by grgrzybek** [comment]
@atooni Thanks for info
I see you're using `<karaf.version>2.3.6</karaf.version>`. Here's Karaf's commit that you should include: https://github.com/apache/karaf/commit/5b29a4466165b519637b9aa9e0b9c3bf0794b7b9. It should be available in next Karaf 2.x version (see https://issues.apache.org/jira/browse/KARAF-4351)

regards
Grzegorz

**Comment by atooni** [comment]
@grgrzybek Thanks for the info. I guess you have seen that we have a karaf version in our parent pom file. This is deprecated as we are not using Karaf as a runtime anymore. We have our own Scala based container framework by now on top of Apache Felix.

**Comment by davsclaus** [comment]
@atooni do you have a ballpark figure of how many queues you have in the AMQ broker, and how many Camel routes are in the same JVM?

**Comment by atooni** [comment]
we have about 50 queues and 50 routes
we do have quite a number of temporary queues in one request reply service

**Comment by davsclaus** [comment]
Okay thanks that isn't to much. So its there alot of other stuff in your JMX tree? Since you can hit that max col size limit ?

**Comment by davsclaus** [comment]
I dont think temporary queues are enlisted in JMX or are they? Can't remember.

**Comment by grgrzybek** [comment]
@atooni try
```
curl 'http://localhost:PORT/hawtio/jolokia/list?maxDepth=2' -H 'Authorization: Basic BASE64_OF_YOUR_LOGIN:PASSWORD' -H 'Connection: close' | jq .
```
you'll see your JMX domains/mbeans.

**Comment by atooni** [comment]
Hi, 

we have executed the query on one of our servers that seems to be "broken". Apparently we have a gazillion MBeans in there that seem to originate from one of our request / reply routes (the clientID is of that route). 

I am attaching the mBeans for information. 
[hawtobjects.zip](https://github.com/hawtio/hawtio/files/150748/hawtobjects.zip)

**Comment by davsclaus** [comment]
ActiveMQ has a lot of options, try to find out if there is an option to not enlist dynamic client consumers.

**Comment by atooni** [comment]
Just inspected the code. I am using a normal Camel producer template to control the request / reply and seem to have thousands of dynamic producer MBeans from this request / reply route:

_1_5_17563,type=Broker":"1","brokerName=sibA,clientId=StoreServices-A-RS,endpoint=dynamicProducer,producerId=ID_de4320lnxprx01-39915-1448894868906-

Need to find out why the producer MBeans don't vanish

**Comment by grgrzybek** [comment]
@atooni hi
I have the fix ready. However, not in hawtio/hawtio repository yet.
Let me do a quick summary of my improvement.
hawtio, when started did these:
* jolokia `LIST` request with `maxDepth: 7` to fetch all available MBeans and their JSONified MBeanInfo. Such big maxDepth (hardcoded) was needed because of RBAC facility provided by Karaf by default (there's dummy version provided by hawtio itself, but in this case, we don't actually need such deep `LIST` request)
* jolokia `SEARCH` request to find MBeans that can provide RBAC methods (Karaf one or dummy one from hawtio)
* add JMX tree postprocessor that prepare bulk RBAC request - very very inefficient in case where you have lots of MBeans created from the same standard MBean (like ActiveMQ queue or any of 12 Camel elements: processors, routes, endpoints, services, contexts, typeconverters, ...)
* when using dummy hawtio implementation, RBAC call wasn't that much of a problem

The problem in your case is actually initial, deep `LIST` request.

My fix brings *special* MBean provided by `hawtio-osgi-jmx` bundle - if it's detected **very early during hawtio client startup**, it's used instead of generic, deep jolokia `LIST` call. Instead hawtio will call this MBean (`hawtio:type=security,name=RBACRegistry`) and will get JSON that is compatible with one returned from generic jolokia `LIST` operation - but this time, we don't repeat the same MBeanInfo all over for each of repeating queues (and Camel elements).

I did a test (2000 ActiveMQ queues in JBoss Fuse) - we wen't down from ~50MB and >1 minute of combined `LIST` + RBAC requests to ... ~500 kB of JSON and ~2 seconds.

I'll see how easy it is to merge these fixes.

I see one problem - now, `hawtio-osgi-jmx` bundle has compile (and Import-Package) dependency on `org.apache.karaf.management.server` bundle...

**Comment by lefou** [comment]
How about making that dependency optional (resolution:=optional) at runtime and put the early startup detection into a separate class (simple anonymous Runnable would be enough) and guard the call to it by a NoClassDefFoundError / ClassNotFoundException try-catch block.

**Comment by lefou** [comment]
@grgrzybek  The problem is, at least in the case of the blended container, that there is no `org.apache.karaf.management.server` bundle running and I read your comment that way, that your fix depends on this bundle. So, the fix would not change @atooni 's situation.

**Comment by grgrzybek** [comment]
@atooni @lefou @davsclaus I've ported the fix to `master` branch now.
I've refactored OSGI-related pieces to `hawtio-osgi-jmx`, but the main improvement (optimized Jolokia `list`) is now in `hawtio-system`. This jar is simply part of `hawtio-web` (and as consequence also in `hawtio-wildfly`).

Now, having only `hawtio-web` WAR we have new JMX bean `hawtio:type=security,name=RBACRegistry`. Hawtio client application verifies the presence of this MBean and possibly switches from generic Jolokia `list` operation to `exec` invocation on this bean. So it can automatically switch to optimized version, where some (whitelisted) cases are optimized - like ActiveMQ queues or Camel elements (routes, endpoints, processors, ...).

Additionaly, when there's another MBean found (`hawtio:type=security,area=jolokia,name=RBACDecorator`), the result of optimized `list` operation may be decorated with RBAC info - this happens in OSGi env when `hawtio-osgi-jmx` bundle is started.

So, @atooni you just have to use new version of `hawtio-web` - no more dependencies on Karaf or `hawtio-osgi-jmx`.
